File: apps/cli-go/internal/utils/supabase_home.go
package utils

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/go-errors/errors"
)

// SupabaseHomeDir returns the global Supabase CLI state root. It is overridden
// by the SUPABASE_HOME environment variable when set to a non-empty value (an
// absolute path is expected; the value is used verbatim), otherwise it defaults
// to ~/.supabase.
func SupabaseHomeDir() (string, error) {
	if home := strings.TrimSpace(os.Getenv("SUPABASE_HOME")); home != "" {
		return home, nil
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return "", errors.Errorf("failed to get $HOME directory: %w", err)
	}
	return filepath.Join(home, ".supabase"), nil
}

File: apps/cli/src/shared/config/supabase-home.ts
import { join } from "node:path";
import { Option } from "effect";

/**
 * Resolves the global Supabase CLI state root.
 *
 * `SUPABASE_HOME` overrides the location when set to a non-empty value after trimming
 * surrounding whitespace (an absolute path is expected; the value is used verbatim). Otherwise it
 * defaults to `<homeDir>/.supabase`. A pure function, so every caller resolves through it with
 * its own environment and home directory, keeping the contract in one place.
 */
export const resolveSupabaseHomeValue = (value: Option.Option<string>, homeDir: string): string => {
  const configured = Option.isSome(value) ? value.value.trim() : undefined;
  return configured !== undefined && configured.length > 0
    ? configured
    : join(homeDir, ".supabase");
};

export const resolveSupabaseHome = (
  env: Readonly<Record<string, string | undefined>>,
  homeDir: string,
): string =>
  resolveSupabaseHomeValue(
    env["SUPABASE_HOME"] === undefined ? Option.none() : Option.some(env["SUPABASE_HOME"]),
    homeDir,
  );
